File: internal/opensse/executor/azure.go
package executor

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strings"
)

func init() {
	RegisterSpecialized("azure", &AzureExecutor{
		Base: Base{
			Provider: "azure",
			Client:   nil,
			Headers:  nil,
			BaseURL:  "",
			BaseURLs: nil,
		},
	})
}

// AzureExecutor executes OpenAI chat completions hosted on Azure.
type AzureExecutor struct{ Base }

func resolveAzureURL(cred Credentials, model string) string {
	endpoint := strPSD(cred, "azureEndpoint")
	if endpoint == "" {
		endpoint = envOr("AZURE_ENDPOINT", "https://api.openai.com")
	}

	apiVersion := strPSD(cred, "apiVersion")
	if apiVersion == "" {
		apiVersion = envOr("AZURE_API_VERSION", "2024-10-01-preview")
	}

	deployment := strPSD(cred, "deployment")
	if deployment == "" {
		deployment = model
	}

	if deployment == "" {
		deployment = envOr("AZURE_DEPLOYMENT", "gpt-4")
	}

	endpoint = strings.TrimRight(endpoint, "/")

	return fmt.Sprintf("%s/openai/deployments/%s/chat/completions?api-version=%s", endpoint, deployment, apiVersion)
}

func buildAzureHeaders(cred Credentials, stream bool) http.Header {
	h := make(http.Header)
	h.Set("Content-Type", "application/json")

	apiKey := cred.APIKey
	if apiKey == "" {
		apiKey = cred.AccessToken
	}

	if apiKey == "" {
		apiKey = os.Getenv("OPENAI_API_KEY")
	}

	if apiKey != "" {
		h.Set("api-key", apiKey)
	}

	if org := strPSD(cred, "organization"); org != "" {
		h.Set("OpenAI-Organization", org)
	}

	if stream {
		h.Set("Accept", "text/event-stream")
	}

	return h
}

// Execute performs Azure OpenAI chat completions.
func (e *AzureExecutor) Execute(ctx context.Context, cred Credentials, model string, body []byte, stream bool) (*Result, error) {
	var m map[string]any
	if err := json.Unmarshal(body, &m); err != nil {
		return nil, err
	}

	m["stream"] = stream

	url := resolveAzureURL(cred, model)

	payload, err := json.Marshal(m)
	if err != nil {
		return nil, err
	}

	h := buildAzureHeaders(cred, stream)

	return e.DoPOST(ctx, url, h, payload)
}

func strPSD(cred Credentials, key string) string {
	if cred.ProviderSpecificData == nil {
		return ""
	}

	if v, ok := cred.ProviderSpecificData[key].(string); ok {
		return v
	}

	return ""
}

func envOr(k, def string) string {
	if v := os.Getenv(k); v != "" {
		return v
	}

	return def
}

File: open-sse/src/executors/azure.rs
//! Azure executor - Custom deployments and headers
use anyhow::{Context, Result};
use reqwest::Client;
use serde_json::Value;
use std::time::Duration;
use crate::executors::default::{UpstreamBody, UpstreamResponse};
use crate::providers::Provider;

pub async fn execute(
    _provider: &Provider,
    model: &str,
    body: Value,
    stream: bool,
    credentials: &Value,
) -> Result<UpstreamResponse> {
    let client = Client::builder()
        .use_rustls_tls()
        .connect_timeout(Duration::from_secs(30))
        .build()?;

    let azure_endpoint = credentials
        .pointer("/providerSpecificData/azureEndpoint")
        .and_then(Value::as_str)
        .map(String::from)
        .or_else(|| std::env::var("AZURE_ENDPOINT").ok())
        .unwrap_or_else(|| "https://api.openai.com".to_string());
    let azure_endpoint = azure_endpoint.trim_end_matches('/');

    let api_version = credentials
        .pointer("/providerSpecificData/apiVersion")
        .and_then(Value::as_str)
        .map(String::from)
        .or_else(|| std::env::var("AZURE_API_VERSION").ok())
        .unwrap_or_else(|| "2024-10-01-preview".to_string());

    let deployment = credentials
        .pointer("/providerSpecificData/deployment")
        .and_then(Value::as_str)
        .map(String::from)
        .or_else(|| if !model.is_empty() { Some(model.to_string()) } else { None })
        .or_else(|| std::env::var("AZURE_DEPLOYMENT").ok())
        .unwrap_or_else(|| "gpt-4".to_string());

    let url = format!("{azure_endpoint}/openai/deployments/{deployment}/chat/completions?api-version={api_version}");

    let mut req = client.post(&url).json(&body);
    if !stream {
        req = req.timeout(Duration::from_secs(300));
    }

    let api_key = credentials
        .get("apiKey")
        .or_else(|| credentials.get("accessToken"))
        .or_else(|| credentials.get("api_key"))
        .and_then(Value::as_str)
        .unwrap_or("");

    if !api_key.is_empty() {
        req = req.header("api-key", api_key);
    }

    if let Some(org) = credentials.pointer("/providerSpecificData/organization").and_then(Value::as_str) {
        req = req.header("OpenAI-Organization", org);
    }

    if stream {
        req = req.header("Accept", "text/event-stream");
    }

    let resp = req.send().await.context("azure request")?;
    let status = resp.status().as_u16();

    if stream && status == 200 {
        use futures_util::StreamExt;
        let byte_stream = resp.bytes_stream();
        let mapped = byte_stream.map(|r| r.map_err(|e| anyhow::anyhow!(e)));
        return Ok(UpstreamResponse {
            status,
            body: UpstreamBody::Sse(Box::pin(mapped)),
            url,
        });
    }

    let json: Value = resp.json().await.context("parse azure json")?;
    Ok(UpstreamResponse {
        status,
        body: UpstreamBody::Json(json),
        url,
    })
}
